runtime/gc_include/ObjectAccessBarrierAPI.hpp

```diff
@@ -357,41 +357,70 @@ class MM_ObjectAccessBarrierAPI
 	VMINLINE void
 	cloneObject(J9VMThread *currentThread, j9object_t original, j9object_t copy, J9Class *objectClass)
 	{
-#if defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) 
-		currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneObject(currentThread, original, copy);
+		copyObjectFields(currentThread, objectClass, original, mixedObjectGetHeaderSize(), copy, mixedObjectGetHeaderSize());
+	}
+
+
+	/**
+	 * Copy valueType from sourceObject to destObject
+	 * See MM_ObjectAccessBarrier::copyValue for detailed description
+	 *
+	 * @param vmThread vmthread token
+	 * @param valueClass The valueType class
+	 * @param srcObject The object being used.
+	 * @param srcOffset The offset of the field.
+	 * @param destObject The object being used.
+	 * @param destOffset The offset of the field.
+	 */
+	VMINLINE void
+	copyObjectFields(J9VMThread *vmThread, J9Class *objectClass, j9object_t srcObject, UDATA srcOffset, j9object_t destObject, UDATA destOffset)
+	{
+#if defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER)
+		vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyObjectFields(vmThread, objectClass, srcObject, srcOffset, destObject, destOffset);
 #else /* defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) */
 		/* TODO implement HW barriers */
 		if (j9gc_modron_readbar_none != _readBarrierType) {	
 			if (j9gc_modron_readbar_evacuate == _readBarrierType) {
-				currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneObject(currentThread, original, copy);
+				vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyObjectFields(vmThread, objectClass, srcObject, srcOffset, destObject, destOffset);
 			} else if (j9gc_modron_readbar_always == _readBarrierType) {
-				currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneObject(currentThread, original, copy);
+				vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyObjectFields(vmThread, objectClass, srcObject, srcOffset, destObject, destOffset);
 			}
 		} else {
-			UDATA offset = mixedObjectGetHeaderSize();
-			UDATA limit = offset + mixedObjectGetDataSize(objectClass, original);
+			UDATA offset = 0;
+			bool isValueType = J9_IS_J9CLASS_VALUETYPE(objectClass);
+			UDATA limit = mixedObjectGetDataSize(objectClass, isValueType);
+			
 			while (offset < limit) {
-				/* No need for pre-store barrier as we are always overwriting 0's in the new object */
-				*(fj9object_t *)((UDATA)copy + offset) = *(fj9object_t *)((UDATA)original + offset);
+				*(fj9object_t *)((UDATA)destObject + offset + destOffset) = *(fj9object_t *)((UDATA)srcObject + offset + srcOffset);
 				offset += sizeof(fj9object_t);
 			}
-		
+			
 #if defined(J9VM_THR_LOCK_NURSERY)
-			/* zero lockword, if present */
-			j9objectmonitor_t *lockwordAddress = getLockwordAddress(copy);
-			if (NULL != lockwordAddress) {
-				*lockwordAddress = 0;
+			if (!isValueType) {
+				/* zero lockword, if present */
+				j9objectmonitor_t *lockwordAddress = getLockwordAddress(destObject);
```

Comment:
hmm for lockword I am wondering if it makes more sense for getLockwordAddress to just return NULL if the object passed in is a ValueType... That might simply things since all callers should be checking NULL now anyway.. It should not be any different at runtime but it should be safer for callers of getLockwordAddress that may not know about ValueTypes.